Create DeepSeek-R1-Tutorial.md (#1215)
microsoft/onnxruntime-genai: examples/python/DeepSeek-R1Tutorial.md

> ## 1. Pre-Requisites: Make venv and install ort genai

Reviewer: ```suggestion
## 1. Pre-Requisites: Make a virtual environment and install ONNX Runtime GenAI
```

```bash
python -m venv .venv && source .venv/bin/activate
pip install requests numpy --pre onnxruntime-genai
```

> 

Reviewer: We should clarify the differences between the top and below instructions (i.e. one is for CPU, one is for CUDA). Users commonly get confused between the different packages.

```bash
python -m venv .venv && source .venv/bin/activate
pip install requests numpy --pre onnxruntime-genai-cuda "olive-ai[gpu]"
```

## 2. Acquire model

```bash
huggingface-cli download onnxruntime/DeepSeek-R1-Distill-ONNX --include 'deepseek-r1-distill-qwen-1.5B/*' --local-dir .
```
OR choose your model and convert to ONNX

```bash
olive auto-opt --model_name_or_path deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B --output_path ./deepseek-r1-distill-qwen-1.5B --device gpu --provider CUDAExecutionProvider --precision int4 --use_model_builder --log_level 1
```

## 3. Play with model

```bash
curl -s https://raw.githubusercontent.com/microsoft/onnxruntime-genai/refs/heads/main/examples/python/model-chat.py

> python model-chat.py -m deepseek-r1-distill-qwen-1.5B -e gpu --chat_template "<|begin▁of▁sentence|><|User|>{input}<|Assistant|>"

Reviewer: ```suggestion
python model-chat.py -m deepseek-r1-distill-qwen-1.5B -e cuda --chat_template "<|begin▁of▁sentence|><|User|>{input}<|Assistant|>"
```

```

> 

Reviewer: Same comment as earlier for distinguishing the examples. A header that clearly states one is for CPU and one is for GPU (e.g. CUDA or DML) would be helpful. [Here](https://github.com/microsoft/onnxruntime-genai/blob/main/examples/python/phi-3.5-vision.md#4-run-phi-35-vision-onnx-models) is an example of the headers.

```bash

> curl -s <https://raw.githubusercontent.com/microsoft/onnxruntime-genai/refs/heads/main/examples/python/model-chat.py>

Reviewer: ```suggestion
curl -s https://raw.githubusercontent.com/microsoft/onnxruntime-genai/refs/heads/main/examples/python/model-chat.py
```

> python model-chat.py -m deepseek-r1-distill-qwen-1.5B/cpu_and_mobile/cpu-int4-rtn-block-32-acc-level-4/ -e cpu

Reviewer: We should add the chat template in the CPU example as well.

## 1. Pre-Requisites: Make a virtual environment and install ONNX Runtime GenAI
```bash
# Installing onnxruntime-genai, olive, and dependencies for CPU

> python -m venv .venv && source .venv/bin/activate

Reviewer: Can we show instructions for a virtual environment using `conda` so that a Python version can be specified? The GenAI package is only available for Python 3.10, 3.11, and 3.12 currently.

pip install requests numpy --pre onnxruntime-genai-cuda "olive-ai[gpu]"
```

## 2. Acquire model

Choose your model and convert to ONNX. Note that many LLMs work, so feel free to try with other models too:

```bash
# Using Olive auto-opt to pull a huggingface model, optimize for CPU, and quantize to INT4 using RTN. 
olive auto-opt --model_name_or_path deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B --output_path ./deepseek-r1-distill-qwen-1.5B --device cpu --provider CPUExecutionProvider --precision int4 --use_model_builder --log_level 1
```

```bash
# Using Olive auto-opt to pull a huggingface model, optimize for CUDA GPUs, and quantize to INT4 using RTN. 
olive auto-opt --model_name_or_path deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B --output_path ./deepseek-r1-distill-qwen-1.5B --device gpu --provider CUDAExecutionProvider --precision int4 --use_model_builder --log_level 1
```

OR download directly using the Huggingface CLI: 

```bash
# Download the model directly using the huggingface cli
huggingface-cli download onnxruntime/DeepSeek-R1-Distill-ONNX --include 'deepseek-r1-distill-qwen-1.5B/*' --local-dir .
```

## 3. Play with your model on device!
```bash

> #CPU Chat inference. If you pulled the model from huggingface, adjust the model directory (-m) accordingly 

Reviewer: ```suggestion
# CPU Chat inference. If you pulled the model from huggingface, adjust the model directory (-m) accordingly 
```

curl -o https://raw.githubusercontent.com/microsoft/onnxruntime-genai/refs/heads/main/examples/python/model-chat.py
python model-chat.py -m deepseek-r1-distill-qwen-1.5B/model -e cpu --chat_template "<|begin▁of▁sentence|><|User|>{input}<|Assistant|>"
```

```bash

> #GPU Chat inference. If you pulled the model from huggingface, adjust the model directory (-m) accordingly 

Reviewer: ```suggestion
# GPU Chat inference. If you pulled the model from huggingface, adjust the model directory (-m) accordingly 
```

curl -o https://raw.githubusercontent.com/microsoft/onnxruntime-genai/refs/heads/main/examples/python/model-chat.py

> python model-chat.py -m deepseek-r1-distill-qwen-1.5B/model -e cpu --chat_template "<|begin▁of▁sentence|><|User|>{input}<|Assistant|>"

Reviewer: ```suggestion
python model-chat.py -m deepseek-r1-distill-qwen-1.5B/model -e cuda --chat_template "<|begin▁of▁sentence|><|User|>{input}<|Assistant|>"
```
